### ISSUE TITLE
Warning while compiling with Clang 4.0 and libc++

### ISSUE THREAD

**Issue by sytelus** [rpclib/rpclib#142]
I get `-Wmismatched-tags` compiler warning when compiling with Clang+Libc++ 4.0 [(binaries are here)](http://releases.llvm.org/4.0.1/clang+llvm-4.0.1-x86_64-linux-gnu-debian8.tar.xz) 

```
[ 18%] Building CXX object rpclib_wrapper/rpclib/CMakeFiles/AirSim-rpclib.dir/lib/rpc/nonstd/optional.cc.o
In file included from /home/shitals/GitHubSrc/AirSim/external/rpclib/lib/rpc/nonstd/optional.cc:1:
/home/shitals/GitHubSrc/AirSim/external/rpclib/include/rpc/nonstd/optional.hpp:1069:1: warning: 'hash' defined as a class template
      here but previously declared as a struct template [-Wmismatched-tags]
class hash< nonstd::optional<T> >
^
/home/shitals/GitHubSrc/AirSim/llvm-build/output/include/c++/v1/type_traits:397:22: note: did you mean class here?
template <class _Tp> struct _LIBCPP_TEMPLATE_VIS hash;
                     ^
1 warning generated.

```

**Comment by sztomi** [comment]
Thanks for reporting this. I did come across this, but thought I fixed it. I'll include the fix in the next release.

**Comment by ahoarau** [comment]
On macOS, I'm even having an error : 
```
/Users/hoarau/dev/rpclib/include/rpc/nonstd/optional.hpp:1070:7: error: explicit specialization of non-template class 'hash'
class hash< nonstd::optional<T> >
      ^   ~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

That is fixable if including `#include <functional>` in this file. Then I get those warnings, but it builds fine and the examples runs OK.

**Comment by sztomi** [comment]
:man_facepalming: I really need that macOS CI sorted out, don't I? Thanks for letting me know!

**Comment by sztomi** [comment]
This has been released in [v2.2.0](https://github.com/rpclib/rpclib/releases/tag/v2.2.0). Please update.
